android/autodispose-android/src/main/java/com/uber/autodispose/android/AutoDisposeAndroidPlugins.java

```diff
@@ -0,0 +1,106 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.autodispose.android;
+
+import android.support.annotation.Nullable;
+import io.reactivex.exceptions.Exceptions;
+import io.reactivex.functions.BooleanSupplier;
+
+/**
+ * Utility class to inject handlers to certain standard AutoDispose Android operations.
+ */
+public final class AutoDisposeAndroidPlugins {
+
+  private AutoDisposeAndroidPlugins() { }
+
+  @Nullable private static volatile BooleanSupplier onCheckMainThread;
+
+  /**
+   * Prevents changing the plugins.
+   */
+  static volatile boolean lockdown;
+
+  /**
+   * Prevents changing the plugins from then on.
+   * <p>
+   * This allows container-like environments to prevent client messing with plugins.
+   */
+  public static void lockdown() {
+    lockdown = true;
+  }
+
+  /**
+   * Returns true if the plugins were locked down.
+   *
+   * @return true if the plugins were locked down
+   */
+  public static boolean isLockdown() {
+    return lockdown;
+  }
+
+  /**
+   * Sets the preferred main thread checker. If not {@code null}, the {@code mainThreadChecker} will
+   * be preferred in all main thread checks in {@link #onCheckMainThread(BooleanSupplier)} calls.
+   * This can be useful for JVM testing environments, where standard Android Looper APIs cannot be
+   * stubbed and thus should be overridden with a custom check.
+   * <p>
+   * This is a reset-able API, which means you can pass {@code null} as the parameter value to reset
+   * it. Alternatively, you can call {@link #reset()}.
+   *
+   * @param mainThreadChecker a {@link BooleanSupplier} to call to check if current execution is
+   * on the main thread. Should return {@code true} if it is on the main
+   * thread or {@code false} if not.
+   */
+  public static void setOnCheckMainThread(@Nullable BooleanSupplier mainThreadChecker) {
+    if (lockdown) {
+      throw new IllegalStateException("Plugins can't be changed anymore");
+    }
+    onCheckMainThread = mainThreadChecker;
+  }
+
+  /**
+   * Returns {@code true} if called on the main thread, {@code false} if not. This will prefer a set
+   * checker via {@link #setOnCheckMainThread(BooleanSupplier)} if one is present, otherwise it will
+   * use {@code defaultChecker}.
+   *
+   * @param defaultChecker the default checker to fall back to if there is no main thread checker
+   * set.
+   * @return {@code true} if called on the main thread, {@code false} if not.
+   */
+  public static boolean onCheckMainThread(BooleanSupplier defaultChecker) {
+    if (defaultChecker == null) {
+      throw new NullPointerException("defaultChecker == null");
```

Comment:
"defaultChecker must not be null" is a better practice for wording exceptions.

Reply:
we follow the rxjava convention for this here and elsewhere in autodispose, would rather discuss/address that in a followup since its a larger scope than this PR